### PR TITLE
memory viz: Add colors for categories and a legend

### PR DESCRIPTION
**Pull request by zdevito**
Adds a category legend to memory trace plots that colors allocations by their role (activation, parameter, gradient, etc.) as captured by kineto.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95753
* #95541
* #95357
* #95356
* #91336
* __->__ #90587

Differential Revision: [D43757381](https://our.internmc.facebook.com/intern/diff/D43757381)

### CODE REVIEW

**Comment by zdevito** [comment]
Screen Shot 2022-12-09 at 4 20 55 PM

File: torch/cuda/_memory_viz.py [hunk 1]
```diff
@@ -430,11 +463,11 @@ def profile_plot(memory_profile, device=None):
     def add_element(size, tensor_key, version):
         category = memory_profile._categories.get(tensor_key, version)
         if category is None:
-            category = 'unknown'
-        else:
-            category = category.name.lower()
+            category = Category.TEMPORARY
```
Comment (robieta) [inline torch/cuda/_memory_viz.py:466]: Why the change? Just because something is uncategorized doesn't mean it's `TEMPORARY`.

Comment (zdevito) [reply torch/cuda/_memory_viz.py:466]: I didn't know what to categorize something without a category. Should we make an unknown or uncategorized category in the legend?

File: test/test_cuda.py [hunk 2]
```diff
@@ -22,6 +22,9 @@
 import torch
 import torch.cuda
 import torch.cuda.comm as comm
+from torch.cuda._memory_viz import profile_plot
+from torch.cuda._memory_viz import trace_plot
```
Comment (aaronenyeshi) [inline test/test_cuda.py:26]: Could you please add some brief comments in the function stubs about what profile_plot and trace_plot are? Also, a brief description of what this PR is changing the PR body would be helpful too.

**Review by aaronenyeshi — CHANGES_REQUESTED** [review]
Overall looks good, would like a little more comments in the _memory_viz.py tool.

File: test/test_cuda.py (hunk 2 above)
Comment (zdevito) [reply test/test_cuda.py:26]: yep, added docstrings and PR comment.

**Review by aaronenyeshi — APPROVED** [review]
Very nice! Thanks for the PR!

**Comment by zdevito** [comment]
@zdevito has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43757381).

**Comment by seemethere** [comment]
I have no idea how this happened but D43757381 is somehow linked to this PR when it should actually be linked to https://github.com/pytorch/pytorch/pull/95541

**Comment by seemethere** [comment]
Just reverted this manually with https://github.com/pytorch/pytorch/commit/b38b39c441f12be90fd5d7eafe74246d050665c8
